enhancements/fbc-run-bundle.md

```diff
@@ -0,0 +1,268 @@
+These three structs are used to construct the declarative config object. This declarative config object is a File-Based Catalog representation of an index with a single bundle present in it. This declarative config object can easily be transformed into `YAML/JSON` format.
+
+It is noteworthy that this enhancement proposal does not change the `run bundle` command from a user’s perspective. The command-line arguments and flags will remain the same. However, the Operator SDK will handle adding a newly-generated bundle to an index differently than it currently does. That is the goal of this enhancement proposal.
+
+#### Command Line Interface
+
+The command line interface will remain exactly the same as the original `run bundle` command.
+
+#### Assumptions
+
+Like the existing run bundle command, we **assume** the following:
+
+ 1. OLM is already installed
+ 1. kube API is available with the permissions to create resources
+ 1. kubelet can see and pull the provided images (i.e. bundle and index images)
+ 1. If an index image is provided, we assume that it has:
+    - bash, so we can change the entrypoint
+    - if it is a sqlite index image, we assume that:
+        - the existing registry DB located at `/database/index.db`
+        - dependencies required for `opm registry add` so we can inject the bundle into the DB
+
+#### What happens when the `run bundle` command is invoked?
+
+1. The labels from the bundle annotations are parsed. The annotations are received from the bundle image that is passed in as a required argument for the run bundle command. Operator SDK will parse information like channel name and CSV name from the labels in the bundle annotations.
+
+2. The Operator SDK, using the data gathered from the above step as parameters, will invoke Operator Registry APIs that will generate three major components in the form of Operator Registry-defined objects - Package, Channel & Bundle.
+
+3. The Operator SDK will convert these objects into more commonly-consumable formats like JSON/YAML blobs and combine these blobs together to generate a basic file-based catalog representation of the bundle image. In other words, the simplest file-based catalog consists of one package blob, one channel blob, and one bundle blob. These blobs represent the bundle data.
+
+4. If an index image was passed in as a parameter through the `--index-image` flag of the run bundle command, a series of additional steps take place:
+
+    - The index image will be rendered into a declarative config object, similar to the bundle image.
+    - Operator SDK will check if the bundle image is already present in the index image.
+        - If the bundle image is already present in the index image, Operator SDK will not add the bundle to the index but rather error out.
+        - If the bundle is not present in the index image, Operator SDK will add the contents of the bundle image to the existing index image and serve the resulting index image (with the bundle added to it).
```

Comment:
Why operator-sdk run bundle allow us to inform an index?
What is the use? I mean, I understand that operator-sdk run bundle will install the bundle and let us check Operator installed via the bundle and using OLM. 

So, what value do we bring by allowing a user to provide an index?
What are the use case scenarios? What SDK run. bundle can do differently according to the index informed?

Shows me that the only extra test/check that we do here when the index is informed is to verify it would be possible add the new bundle to a pre-existent index. Am I right? If yes, should it really be part of the run bundle scope and checks? Should not it only be used/useful for the upgrade bundle command?

Reply:
@camilamacedo86, part of the bundle installation process requires serving a File-Based Catalog representation of the bundle to gRPC via the ephemeral pod. For this reason, we are creating a FBC out of a single bundle (for the default case). If an index image is provided (``--index-image``), then we will add the bundle to this existing index image. If it is not provided, we will create the FBC out of just one bundle image.

Reply:
Sorry, I still have a hard time understanding why provide the option to inform the index at all
what is the goal with? what value does it bring to the purpose of this feature?

Reply:
If the bundle you want to run has dependencies, this enables someone to run `run bundle` and add the bundle to a catalog that also contains that bundle's dependencies.

This can be used to simulate "what happens when I add the next version of my bundle to the existing production catalog"

Reply:
@joelanford, so if the index populated matters because we want to check if is possible to add the new bundle on the index and still run it then, would not the channel used make difference? If yes, how authors will define the channel where the new bundle should be added?

Reply:
Since this PE is merged I am trying to cross the point and shows the gap here in : https://github.com/operator-framework/operator-sdk/pull/5809#discussion_r882810138

IMO the goal of the command checks an upgrade ( upgrade graph ) and not deps so this option leads to a misunderstanding and brings complexities when FBC shows that we are unable to do that.

I am as a user by looking "bundle-upgrade --index" I would not assume that it does not validate the upgrade path ignoring the channels informed in the bundle and it is only useful for checking deps instead of an upgrade graph. So:

- Is it technically possible/achievable that we have a command that validates the upgrade with a previous index with FBC? OR what is achievable?
- How much value vs effort from what is achieved do with a previous index now we have to keep the --index image?

Reply:
My motivation is to try to avoid future scenarios like:
a) Users and pipelines adopting the command and expecting that it is checking the upgrade graph with the data in the index informed ( when it is not ) - false positives
b) we spend effort in issues and troubleshooting it until we remember or figure out that cannot work + or spending time answering questions about.